### PR TITLE
Add: Text ellipsis component

### PR DESCRIPTION
**Pull request by lightbringer1991**
Add TextEllipsis component
textellipsis

### CODE REVIEW

**Comment by lightbringer1991** [comment]
[Add Publishers and Sites modal needs ellipsis](https://trello.com/c/au5NfDzd/6395-add-publishers-and-sites-modal-needs-ellipsis)

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      >
+        {text}
+      </div>
+    );
+  }
+}
+
+TextEllipsisComponent.propTypes = {
+  text: PropTypes.string,
+  overlayTriggerProps: PropTypes.object,
```
Comment (omgaz) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: `overlayTriggerProps: OverlayTrigger.propTypes,`

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  tooltipProps: PropTypes.object,
```
Comment (omgaz) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: `tooltipProps: Popover.propTypes,`

Comment (omgaz) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: also, all of these are required.

File: test/components/adslotUi/TextEllipsisComponentTest.jsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  let divContainer = null;
+
+  beforeEach(() => {
+    divContainer = document.createElement('div');
+    document.body.innerHTML = '';
+    document.body.appendChild(divContainer);
+  });
+
+  it('should render with defaults', () => {
+    const component = mount(<TextEllipsisComponent text="Lorem ipsum dolor sit amet, consectetuer adipiscing elit." />);
```
Comment (omgaz) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: Why are you mounting? We should only need to use shallow rendering.

Comment (omgaz) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: The other's are fine, they allow you to use attachTo

Comment (lightbringer1991) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: `shallow` doesn't render default props. I just want to test the default props. If it's not required, I can remove the test altogether.

Comment (patrick-ivers) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: Really? Are you sure shallow doesn't render default props?

Comment (lightbringer1991) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: @patrick-ivers yea it returns undefined for props that should have default values.

Comment (patrick-ivers) [inline test/components/adslotUi/TextEllipsisComponentTest.jsx]: Instead of using `mount()`, can you try with `shallowWrapper.instance().props()`?
[http://airbnb.io/enzyme/docs/api/ShallowWrapper/props.html](props())
[http://airbnb.io/enzyme/docs/api/ShallowWrapper/instance.html](instance())

File: README.md [hunk 4]
```diff
@@ -107,3 +107,17 @@ window.Perf = Perf;
...
+- Commits have to be squashed into 1 commit per PR
+- Commit message has to be prefixed with:
+  - 'Fix:'
+  - 'Update:'
+  - 'Breaking:'
+  - 'Docs:'
+  - 'Build:'
+  - 'New:'
+  - 'Upgrade:'
+  - 'Release:'
```
Comment (omgaz) [inline README.md]: Seems like this should be in contributing.md and not readme

Comment (lightbringer1991) [inline README.md]: ok I'll move it over

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      truncated: false,
+    };
+  }
+
+  componentDidMount() {
+    // determine if the text is truncated or not
+    if (this.container.scrollWidth > this.container.clientWidth) {
```
Comment (xiaofan2406) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: it wouldn't change when `this.props.text` is change. So in the case if initially `text` is long and needs to truncated, it sets the state correctly. But when the `text` changed from parent to something really short, it will stay truncated.

File: src/components/Main.jsx [hunk 6]
```diff
@@ -982,6 +983,12 @@ class AppComponent extends React.Component {
           ]}
         />
 
+        <h2>TextEllipsis</h2>
```
Comment (xiaofan2406) [inline src/components/Main.jsx]: Can we add demo into its seperate example? and imported in the component?

Comment (lightbringer1991) [inline src/components/Main.jsx]: the demo is only in `Main.jsx` though, I'm not sure if we need to split demos into multiple files

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+  render() {
+    const {
+      text,
+      popoverProps,
+      overlayTriggerProps,
+    } = this.props;
+    const { truncated } = this.state;
+
+    const tooltip = (
```
Comment (xiaofan2406) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: this should be inside the if statement below

File: src/components/Main.jsx (hunk 6 above)
Comment (xiaofan2406) [inline src/components/Main.jsx]: the `AppComponent` is way too big. We should seperate new demo into its own component as we touch them. So it doesn't rerender everything on one change.

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return ((this.state.truncated !== nextState.truncated) || (this.props.text !== nextProps.text));
+  }
+
+  componentDidUpdate() {
+    if (this.container.scrollWidth > this.container.clientWidth) {
+      this.setTruncate(true);
```
Comment (xiaofan2406) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: this.setTruncate(this.container.scrollWidth > this.container.clientWidth)

File: src/components/adslotUi/TextEllipsisComponent.jsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    } else {
+      this.setTruncate(false);
+    }
+  }
+
+  setTruncate(isTruncated) {
```
Comment (xiaofan2406) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: this should still be binded even though the current usage isn't causing problem

Comment (sonhanguyen) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: can you get rid of the param and put the block `if (this.container.scrollWidth > this.container.clientWidth) {` so that it's not duplicated?

File: src/components/Main.jsx (hunk 6 above)
Comment (patrick-ivers) [inline src/components/Main.jsx]: @xiaofan2406 Let's do this in a separate PR.

**Comment by patrick-ivers** [comment]
@lightbringer1991 Could you please add a description to this PR.

File: src/components/Main.jsx (hunk 6 above)
Comment (xiaofan2406) [inline src/components/Main.jsx]: @patrick-ivers We should probably put new examples in the examples folder as adding them, so it's less work for the refactor later.

File: src/components/adslotUi/TextEllipsisComponent.jsx (hunk 9 above)
Comment (sonhanguyen) [inline src/components/adslotUi/TextEllipsisComponent.jsx]: @xiaofan2406 to me prototype method is pointless because you end up creating a bound one anyway. I prefer just put everything in a constructor so that you don't have to create two functions for the same thing. I'm not sure if it makes calling constructor more expensive (may not because JIT), but whatever the cost is we gain back from adding those to prototype at load time.

**Comment by patrick-ivers** [comment]
Merging...

File: src/components/Main.jsx (hunk 6 above)
Comment (patrick-ivers) [inline src/components/Main.jsx]: @xiaofan2406 Actually, we want to possibly go a different way with this...

```
components/
├── MyComponent/
│   ├── component.jsx
│   ├── component.spec.jsx
│   ├── component.example.jsx
│   └── styles.scss
```

Comment (sonhanguyen) [inline src/components/Main.jsx]: @omgaz @patrick-ivers we can even get rid of `Main.jsx` and let something like `react-storybook` gather them into one place: https://github.com/storybooks/storybook/blob/master/examples/cra-storybook/.storybook/config.js
